saleor/graphql/core/connection.py

```diff
@@ -1,7 +1,180 @@
+    if isinstance(last, int) and last < 0 or last and not isinstance(last, int):
+        raise ValueError("Argument 'last' must be a non-negative integer.")
+    if first and last:
+        raise ValueError("Argument 'last' cannot be combined with 'first'.")
+    if first and args.get("before"):
+        raise ValueError("Argument 'first' cannot be combined with 'before'.")
+    if last and args.get("after"):
+        raise ValueError("Argument 'last' cannot be combined with 'after'.")
+
+
+def connection_from_queryset_slice(
+    qs: QuerySet,
+    args: ConnectionArguments = None,
+    connection_type: typing.Any = Connection,
+    edge_type: typing.Any = Edge,
+    page_info_type: typing.Any = PageInfo,
+) -> Connection:
+    """Create a connection object from a QuerySet."""
+    args = args or {}
+    before = args.get("before")
+    after = args.get("after")
+    first = args.get("first")
+    last = args.get("last")
+    connection_args_validation(args)
+
+    cursor_after = from_global_cursor(after) if after else None
+    cursor_before = from_global_cursor(before) if before else None
+    cursor = cursor_after or cursor_before
+    cursor_offset = 1 if cursor else 0
+
+    sort_by = args.get("sort_by", {})
+    sorting_direction = sort_by.get("direction", "")
+    sorting_fields = sort_by.get("field")
+    if sorting_fields and not isinstance(sorting_fields, list):
+        sorting_fields = [sorting_fields]
+    elif not sorting_fields:
+        sorting_fields = []
+
+    requested_count = first or last
+    end_margin = requested_count + cursor_offset if requested_count else None
+
+    if last:
+        # reversed direction
+        sorting_direction = "gt" if sorting_direction == OrderDirection.DESC else "lt"
+    else:
+        sorting_direction = "lt" if sorting_direction == OrderDirection.DESC else "gt"
+
+    filter_kwargs = (
+        prepare_filter(cursor, sorting_fields, sorting_direction) if cursor else Q()
+    )
+    qs = qs.filter(filter_kwargs)[:end_margin]
+    if last:
+        matching_records = list(reversed(qs))
+    else:
+        matching_records = list(qs)
+    has_previous_page = False
+    has_next_page = False
+    if requested_count is not None:
+        if cursor_after:
+            has_next_page = len(matching_records) > requested_count
+            has_previous_page = True
+        elif cursor_before:
+            has_next_page = True
+            has_previous_page = len(matching_records) > requested_count
+        elif first:
+            has_next_page = len(matching_records) > requested_count
+        elif last:
+            has_previous_page = len(matching_records) > requested_count
+        matching_records = matching_records[:requested_count]
+
+    edges = [
+        edge_type(
+            node=record,
+            cursor=to_global_cursor(
+                [get_field_value(record, field) for field in sorting_fields]
+            ),
+        )
+        for index, record in enumerate(matching_records)
```

Comment:
Excuse me passing by...
looking at commit history, `index` is not needed anymore
also why do we need to calculate cursor for every edge? (is there any spec?)

Reply:
The spec is here: https://facebook.github.io/relay/graphql/connections.htm